rcl_yaml_param_parser/src/parse.c

```diff
@@ -0,0 +1,832 @@
+    style != YAML_DOUBLE_QUOTED_SCALAR_STYLE)
+  {
+    errno = 0;
+    endptr = NULL;
+    dval = strtod(value, &endptr);
+    if ((0 == errno) && (NULL != endptr)) {
+      if ((NULL != endptr) && (endptr != value)) {
+        if (('\0' != *value) && ('\0' == *endptr)) {
+          *val_type = DATA_TYPE_DOUBLE;
+          ret_val = allocator.zero_allocate(1U, sizeof(double), allocator.state);
+          if (NULL == ret_val) {
+            return NULL;
+          }
+          *((double *)ret_val) = dval;
+          return ret_val;
+        }
+      }
+    }
+    errno = 0;
+  }
+
+  /// It is a string
+  *val_type = DATA_TYPE_STRING;
+  ret_val = rcutils_strdup(value, allocator);
+  return ret_val;
+}
+
+///
+/// Parse the value part of the <key:value> pair
+///
+rcutils_ret_t parse_value(
+  const yaml_event_t event,
+  const bool is_seq,
+  const size_t node_idx,
+  const size_t parameter_idx,
+  data_types_t * seq_data_type,
+  rcl_params_t * params_st)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(seq_data_type, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(params_st, RCUTILS_RET_INVALID_ARGUMENT);
+
+  rcutils_allocator_t allocator = params_st->allocator;
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(
+    &allocator, "invalid allocator", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (0U == params_st->num_nodes) {
+    RCUTILS_SET_ERROR_MSG("No node to update");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  const size_t val_size = event.data.scalar.length;
+  const char * value = (char *)event.data.scalar.value;
+  yaml_scalar_style_t style = event.data.scalar.style;
+  const uint32_t line_num = ((uint32_t)(event.start_mark.line) + 1U);
+
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    value, "event argument has no value", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (style != YAML_SINGLE_QUOTED_SCALAR_STYLE &&
+    style != YAML_DOUBLE_QUOTED_SCALAR_STYLE &&
+    0U == val_size)
+  {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("No value at line %d", line_num);
+    return RCUTILS_RET_ERROR;
+  }
+
+  if (NULL == params_st->params[node_idx].parameter_values) {
+    RCUTILS_SET_ERROR_MSG("Internal error: Invalid mem");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
+  rcl_variant_t * param_value = &(params_st->params[node_idx].parameter_values[parameter_idx]);
+
+  // param_value->string_value = rcutils_strdup(value, allocator);
+  data_types_t val_type;
+  void * ret_val = get_value(value, style, &val_type, allocator);
+  if (NULL == ret_val) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error parsing value %s at line %d", value, line_num);
+    return RCUTILS_RET_ERROR;
+  }
+
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+  switch (val_type) {
+    case DATA_TYPE_UNKNOWN:
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Unknown data type of value %s at line %d\n", value, line_num);
+      ret = RCUTILS_RET_ERROR;
+      break;
+    case DATA_TYPE_BOOL:
+      if (false == is_seq) {
```

Comment:
@brawner probably preexistent, but:
```suggestion
      if (!is_seq) {
```
? Same below.

Reply:
Definitely preexistent. Switched